File: src/go/problems/is_path_crossing_test.go
// https://leetcode.com/problems/path-crossing/

package problems

import (
	"testing"

	"github.com/dobermanch/leetcode/core"
)

type IsPathCrossing struct{}

func TestIsPathCrossing(t *testing.T) {
	gen := core.TestSuite[IsPathCrossing]{}
	gen.Add(func(tc *core.TestCase) {
		tc.Param("NES").Result(false)
	}).Add(func(tc *core.TestCase) {
		tc.Param("NESWW").Result(true)
	}).Run(t)
}

func (IsPathCrossing) Solution(path string) bool {
	point := [2]int{0, 0}

	visited := make(map[[2]int]bool)
	visited[point] = true

	for _, direction := range path {
		switch direction {
		case 'N':
			point[1]++
		case 'S':
			point[1]--
		case 'E':
			point[0]++
		case 'W':
			point[0]--
		}

		if visited[point] {
			return true
		}

		visited[point] = true
	}

	return false
}

File: src/python/problems/is_path_crossing_test.py
# https://leetcode.com/problems/path-crossing/

from core.problem_base import *

class IsPathCrossing(ProblemBase):
    def Solution(self, path: str) -> bool:
        point = (0, 0)

        visited = set()
        visited.add(point)

        for direction in path:
            if direction == 'N':
                point = (point[0], point[1] + 1)
            elif direction == 'S':
                point = (point[0], point[1] - 1)
            elif direction == 'E':
                point = (point[0] + 1, point[1])
            elif direction == 'W':
                point = (point[0] - 1, point[1])

            if point in visited:
                return True
            visited.add(point)

        return False

if __name__ == '__main__':
    TestGen(IsPathCrossing) \
        .Add(lambda tc: tc.Param("NES").Result(False)) \
        .Add(lambda tc: tc.Param("NESWW").Result(True)) \
        .Run()
